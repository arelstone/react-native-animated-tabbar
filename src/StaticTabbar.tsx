import React, { FC, Fragment, ReactElement, useState } from 'react';
import { View, TouchableWithoutFeedback, StyleSheet, Animated, StyleProp, TextStyle, Dimensions } from 'react-native';
import Circle, { ActiveCircleProps } from './Circle';
import Label from './Label';
import Constants from './constants';

type Tab = IconProp & ActiveIconProp & {
    label?: string;
    onPress: () => void;
}

export type IconProp = {
    icon: ReactElement;
}

export type ActiveIconProp = {
    activeIcon: ReactElement;
}

export type TabsProp = {
    tabs: Tab[];
    textStyle?: StyleProp<TextStyle>;
    height?: number;
    circleAnimationDurationMs?: number;
}

export type ColorProps = {
    backgroundColor: string;
}
interface StaticTabbarProps extends TabsProp, ActiveCircleProps, ColorProps {
    translateX: Animated.Value;
}

const { width } = Dimensions.get('window');

const StaticTabbar: FC<StaticTabbarProps> = ({
    tabs, translateX, circleStyle, circleContainerStyle, backgroundColor, textStyle,
    circleSize = Constants.circleSize, height = Constants.tabHeight,
    circleAnimationDurationMs = Constants.circleAnimationDurationMs,
}) => {
    const [values] = useState<Animated.Value[]>(tabs.map((_, index) => new Animated.Value(index === 0 ? 1 : 0)));
    const tabWidth = width / tabs.length;
    const handleOnPress = (toIndex: number) => {
        tabs[toIndex].onPress();
        Animated.sequence([
            ...values.map(value => Animated.timing(
                value,
                { toValue: 0, useNativeDriver: true, duration: circleAnimationDurationMs },
            )),
            Animated.parallel([
                Animated.spring(values[toIndex], { toValue: 1, useNativeDriver: true }),
                Animated.spring(translateX, { toValue: -width + tabWidth * toIndex, useNativeDriver: true }),
            ]),
        ]).start();
    };

    return <View style={styles.container}>
        {tabs.map(({ icon, activeIcon, label }, index) => {
            const activeValue = values[index];
            const opacity = translateX.interpolate({
                inputRange: [
                    -width + tabWidth * (index - 1),
                    -width + tabWidth * index,
                    -width + tabWidth * (index + 1),
                ],
                outputRange: [1, 0, 1],
                extrapolate: 'clamp',
            });
            const translateY = activeValue.interpolate({ inputRange: [0, 1], outputRange: [height, 0] });

            return <Fragment key={`Tab_${index}`}>
                <TouchableWithoutFeedback onPress={() => handleOnPress(index)}>
                    <Animated.View style={[styles.tab, { opacity, height }]}>
                        {icon}
                        {label && <Label
                            {...{ label, textStyle }}
                            color={icon.props.color}
                        />}
                    </Animated.View>
                </TouchableWithoutFeedback>
                <Circle
                    {...{
                        activeIcon, backgroundColor, height, index, circleContainerStyle,
                        circleSize, circleStyle, translateY,
                    }}
                    width={tabWidth}
                />
            </Fragment>;
        })}
    </View>;
};

export default StaticTabbar;



const styles = StyleSheet.create({
    container: {
        flexDirection: 'row',
    },
    tab: {
        flex: 1,
        justifyContent: 'center',
        alignItems: 'center',
    },
});
