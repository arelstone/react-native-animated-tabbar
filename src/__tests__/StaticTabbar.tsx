import React from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import StaticTabbar from '../StaticTabbar';
import { Animated, Dimensions, Text } from 'react-native';
import { findElementByTestId } from '../utils/testHelpers';

let wrapper: ShallowWrapper;
const mockOnPress = jest.fn();
const tabs = [
    {
        icon: <Text>Icon #1</Text>,
        activeIcon: <Text>ActiveIcon #1</Text>,
        onPress: mockOnPress,
    },
    {
        icon: <Text>Icon #2</Text>,
        activeIcon: <Text>ActiveIcon #2</Text>,
        onPress: mockOnPress,
    },
];

describe('<StaticTabbar />', () => {
    beforeAll(() => {
        wrapper = shallow(<StaticTabbar
            backgroundColor="blue"
            {...{ tabs }}
            translateX={new Animated.Value(0)}
        />);
    });

    it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    it('should display TWO tabs', () => {
        expect(wrapper.find('TouchableWithoutFeedback')).toHaveLength(2);
    });

    it('should display correct Icon', () => {
        expect(findElementByTestId('TouchableWithoutFeedback', 'Tab_0', wrapper).find('Text').prop('children')).toEqual('Icon #1');
    });

    describe('Circle', () => {
        it('should display a Circle', () => {
            expect(findElementByTestId('Circle', 'Circle_0', wrapper).exists()).toBeTruthy();
        });

        it('should have correct ActiveIcon', () => {
            expect(findElementByTestId('Circle', 'Circle_0', wrapper).prop('activeIcon').props.children).toEqual('ActiveIcon #1');
        });

        it('should apply circleSize', () => {
            wrapper.setProps({ circleSize: 60 });
            expect(findElementByTestId('Circle', 'Circle_0', wrapper).prop('circleSize')).toEqual(60);
        });

        it('should apply height', () => {
            expect(findElementByTestId('Circle', 'Circle_0', wrapper).prop('height')).toEqual(64);
        });

        it('should have correct width', () => {
            expect(findElementByTestId('Circle', 'Circle_0', wrapper).prop('width')).toEqual(Dimensions.get('window').width / tabs.length);
        });
    });
});
