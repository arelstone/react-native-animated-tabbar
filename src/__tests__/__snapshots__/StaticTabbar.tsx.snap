// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<StaticTabbar /> should match snapshot 1`] = `ShallowWrapper {}`;
